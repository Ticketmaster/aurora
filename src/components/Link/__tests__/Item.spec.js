/* global document */
import React from "react";
import { render, Simulate, fireEvent } from "react-testing-library";

import LinkList from "../List";
import LinkItem from "../Item";

const mockFn = jest.fn();

const countries = [
  {
    value: "en-US",
    text: "USA (English)"
  },
  {
    value: "en-CA",
    text: "Canada (English)"
  },
  {
    value: "fr-CA",
    text: "Canada (Fran√ßais)"
  }
];

describe("<LinkItem />", () => {
  const getContainer = props =>
    render(
      <LinkItem role="button" {...props}>
        Button text
        <LinkList onItemClick={mockFn}>
          {countries.map((option, index) => (
            <LinkItem index={index} key={option.value} href="/" role="link">
              {option.text}
            </LinkItem>
          ))}
        </LinkList>
      </LinkItem>
    );

  it("renders LinkItem as button ", () => {
    const { container } = render(
      <LinkItem role="button">Button text</LinkItem>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders LinkItem as link", () => {
    const { container } = render(
      <LinkItem role="link" target="_blank" href="http://localhost/new/">
        Link text
      </LinkItem>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders LinkItem children correctly", () => {
    const { container } = getContainer();
    expect(container).toMatchSnapshot();
  });

  it("should show LinkItem children on click LinkItem if children are hidden and toggleChildrenOnClick passed", () => {
    const { container } = getContainer({ toggleChildrenOnClick: true });
    Simulate.click(container.querySelector(".list-container"));
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should hide LinkItem children on LinkItem click if children are shown and toggleChildrenOnClick passed", () => {
    const { container } = getContainer({ toggleChildrenOnClick: true });
    Simulate.click(container.querySelector(".list-container"));
    Simulate.click(container.querySelector(".list-container"));
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should show LinkItem children on mouse enter", () => {
    const { container } = getContainer();
    Simulate.mouseEnter(container.querySelector(".list-container"));
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should hide LinkItem children on document click if children are shown", () => {
    const { container } = getContainer({ toggleChildrenOnClick: true });
    Simulate.click(container.querySelector(".list-container"));
    fireEvent.click(document);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe("should add/remove document event listener", () => {
    const addEventListener = jest.fn();
    const removeEventListener = jest.fn();
    beforeEach(() => {
      addEventListener.mockClear();
      removeEventListener.mockClear();
      Object.defineProperty(document, "addEventListener", {
        writable: true,
        value: addEventListener
      });
      Object.defineProperty(document, "removeEventListener", {
        writable: true,
        value: removeEventListener
      });
    });

    it("subscribes on document click event on mount", () => {
      const { unmount } = render(
        <LinkItem role="button" toggleChildrenOnClick>
          Button text
        </LinkItem>
      );
      expect(addEventListener).toHaveBeenLastCalledWith(
        "click",
        expect.anything()
      );
      unmount();
    });

    it("unsubscribes from document click event on unmount", () => {
      const { unmount } = render(
        <LinkItem role="button" toggleChildrenOnClick>
          Button text
        </LinkItem>
      );
      unmount();
      expect(addEventListener).toHaveBeenLastCalledWith(
        "click",
        expect.anything()
      );
    });
  });
});
