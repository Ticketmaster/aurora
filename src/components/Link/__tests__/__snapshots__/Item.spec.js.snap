// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinkItem /> renders LinkItem as button  1`] = `
<div>
  <span
    class="list-container list-container--closed"
    role="none"
  >
    <button
      class="link sc-bxivhb fBqTYi"
      role="button"
    >
      Button text
    </button>
  </span>
</div>
`;

exports[`<LinkItem /> renders LinkItem as link 1`] = `
<div>
  <span
    class="list-container list-container--closed"
    role="none"
  >
    <a
      class="link sc-bxivhb fBqTYi"
      href="http://localhost/new/"
      rel="noopener"
      role="link"
      target="_blank"
    >
      Link text
    </a>
  </span>
</div>
`;

exports[`<LinkItem /> renders LinkItem children correctly 1`] = `
<div>
  <span
    class="list-container list-container--closed"
    role="none"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="link link--has-other sc-bxivhb fBqTYi"
      role="button"
    >
      Button text
    </button>
    <div
      class="sc-bdVaJa links links__list bhMvGM"
    >
      <div
        class="sc-bwzfXH bQTlgY"
      >
        <span
          class="list-container list-container--closed"
          role="none"
        >
          <a
            class="link sc-bxivhb fBqTYi"
            href="/"
            role="link"
            target="_self"
          >
            USA (English)
          </a>
        </span>
        <span
          class="list-container list-container--closed"
          role="none"
        >
          <a
            class="link sc-bxivhb fBqTYi"
            href="/"
            role="link"
            target="_self"
          >
            Canada (English)
          </a>
        </span>
        <span
          class="list-container list-container--closed"
          role="none"
        >
          <a
            class="link sc-bxivhb fBqTYi"
            href="/"
            role="link"
            target="_self"
          >
            Canada (Fran√ßais)
          </a>
        </span>
      </div>
    </div>
  </span>
</div>
`;
